tpke/src/refresh.rs

```diff
@@ -0,0 +1,164 @@
+use crate::{lagrange_basis_at, PrivateDecryptionContextSimple};
+use ark_ec::{PairingEngine, ProjectiveCurve};
+use ark_ff::{One, PrimeField, Zero};
+use ark_std::UniformRand;
+use itertools::zip_eq;
+use rand::prelude::StdRng;
+use rand_core::RngCore;
+use std::collections::HashMap;
+use std::usize;
+
+pub fn recover_share_at_point<E: PairingEngine>(
+    other_participants: &[PrivateDecryptionContextSimple<E>],
+    threshold: usize,
+    x_r: &E::Fr,
+    rng: &mut StdRng,
+) -> E::G2Projective {
+    let share_updates = prepare_share_updates_for_recovery::<E>(
+        other_participants,
+        x_r,
+        threshold,
+        rng,
+    );
+
+    let new_shares_y =
+        update_shares_for_recovery::<E>(other_participants, &share_updates);
+
+    // Interpolate new shares to recover y_r
+    let shares_x = &other_participants[0]
+        .public_decryption_contexts
+        .iter()
+        .map(|ctxt| ctxt.domain)
+        .collect::<Vec<_>>();
+
+    // Recover y_r
+    let lagrange = lagrange_basis_at::<E>(shares_x, x_r);
+    let prods =
+        zip_eq(new_shares_y, lagrange).map(|(y_j, l)| y_j.mul(l.into_repr()));
+    prods.fold(E::G2Projective::zero(), |acc, y_j| acc + y_j)
+}
+
+fn prepare_share_updates_for_recovery<E: PairingEngine>(
+    participants: &[PrivateDecryptionContextSimple<E>],
+    x_r: &E::Fr,
+    threshold: usize,
+    rng: &mut impl RngCore,
+) -> HashMap<usize, HashMap<usize, E::G2Projective>> {
+    // TODO: Refactor this function so that each participant performs it individually
+    // Each participant prepares an update for each other participant
+    participants
+        .iter()
+        .map(|p1| {
+            let i = p1.index;
+            // Generate a new random polynomial with constant term 0
+            let d_i = make_random_polynomial::<E>(threshold, x_r, rng);
+
+            // Now, we need to evaluate the polynomial at each of participants' indices
+            let deltas_i: HashMap<_, _> =
+                compute_polynomial_deltas::<E>(participants, &d_i);
+            (i, deltas_i)
+        })
+        .collect::<HashMap<_, _>>()
+}
+
+fn update_shares_for_recovery<E: PairingEngine>(
+    participants: &[PrivateDecryptionContextSimple<E>],
+    deltas: &HashMap<usize, HashMap<usize, E::G2Projective>>,
+) -> Vec<E::G2Projective> {
+    // TODO: Refactor this function so that each participant performs it individually
+    participants
+        .iter()
+        .map(|p| {
+            let i = p.index;
+            let mut new_y = E::G2Projective::from(
+                p.private_key_share.private_key_shares[0], // y_i
+            );
```

Comment:
Why they need to be converted into Projective form?

Reply:
In `arkworks`, `Projective` form is preferred since it "[is generally more efficient for arithmetic](https://docs.rs/ark-algebra-intro/latest/ark_algebra_intro/#elliptic-curve-arithmetic)". The `PrivateKeyShares` are `G2Affine` and so we convert to it [later on](https://github.com/nucypher/ferveo/pull/26/files/8f6f1d8a137440151a71c63bfd33e1ede257e918#diff-b9683cc455c05bd6e307e7a9002573f586accc82b58602db05e96c65e150cd1dR557).